DPPP/test/unit/tMSBDAReader.cc

```diff
@@ -0,0 +1,149 @@
+  MSBDAReader reader(kMSName, parset, prefix);
+
+  reader.setInfo(info);
+
+  info = reader.getInfo();
+  BOOST_TEST(reader.table().tableName().compare(
+                 reader.table().tableName().length() - kMSName.length(),
+                 kMSName.length(), kMSName) == 0);
+  BOOST_TEST(reader.spectralWindow() == -1U);
+  BOOST_TEST(info.nchan() == 16U);
+  BOOST_TEST(info.ncorr() == 4U);
+  BOOST_TEST(info.ntime() == 1U);
+  BOOST_TEST(info.channelsAreRegular());
+  BOOST_TEST(info.hasBDAChannels());
+  BOOST_TEST(info.nbaselines() == 6U);
+  BOOST_TEST(info.startchan() == 0U);
+  BOOST_TEST(info.startTime() == 0U);
+  BOOST_TEST(info.timeInterval() == 30U);
+}
+
+BOOST_AUTO_TEST_CASE(process, *boost::unit_test::tolerance(0.001) *
+                                  boost::unit_test::tolerance(0.0001f)) {
+  DPInfo info;
+  ParameterSet parset;
+  parset.add("memoryperc", "100");
+  parset.add("memory_max", "1");
+  MSBDAReader reader("tNDPPP_bda_tmp.MS", parset, prefix);
+  reader.setReadVisData(true);
+  auto mock_step = std::make_shared<DP3::DPPP::MockStep>();
+  reader.setNextStep(mock_step);
+  reader.setInfo(info);
+  DP3::DPPP::DPBuffer buf;
+
+  reader.process(buf);
+  reader.finish();
+
+  auto kExpectedData = std::complex<float>(2.75794, 0.899097);
+  double kExpectedUVW[] = {-0.125726, -766.917, 397.793};
+  auto kExpectedWeights = std::vector<float>(
+      reader.getInfo().ncorr() * reader.getInfo().nchan(), 1);
+
+  mock_step->CheckFinishCount(1);
+  BOOST_TEST(mock_step->GetBdaBuffers().size() == 1U);
+  BOOST_TEST(mock_step->GetBdaBuffers()[0]->GetRows().size() == 120U);
+  BOOST_TEST(mock_step->GetBdaBuffers()[0]->GetRows()[0].data->imag() ==
+             kExpectedData.imag());
+  BOOST_TEST(mock_step->GetBdaBuffers()[0]->GetRows()[0].data->real() ==
+             kExpectedData.real());
+  BOOST_TEST(mock_step->GetBdaBuffers()[0]->GetRows()[0].uvw == kExpectedUVW);
+  for (unsigned i = 0; i < kExpectedWeights.size(); ++i) {
+    BOOST_TEST(*(mock_step->GetBdaBuffers()[0]->GetRows()[0].weights + i) ==
+               kExpectedWeights[i]);
+  }
+
+  // Check that the print methods do not throw any error
+  std::ostream nullout(nullptr);
+  BOOST_CHECK_NO_THROW(reader.show(nullout));
+  BOOST_CHECK_NO_THROW(reader.showCounts(nullout));
+  BOOST_CHECK_NO_THROW(reader.showTimings(nullout, 10));
+}
+
+BOOST_AUTO_TEST_CASE(process_nan) {
+  DPInfo info;
+  ParameterSet parset;
+  parset.add("memoryperc", "100");
+  parset.add("memory_max", "1");
+  MSBDAReader reader("tNDPPP_bda_tmp.MS", parset, prefix);
+  auto mock_step = std::make_shared<DP3::DPPP::MockStep>();
+  reader.setNextStep(mock_step);
+  reader.setInfo(info);
+  DP3::DPPP::DPBuffer buf;
+
+  reader.setReadVisData(false);
+  reader.process(buf);
+  reader.finish();
+
+  BOOST_TEST(
+      std::isnan(mock_step->GetBdaBuffers()[0]->GetRows()[0].data->imag()));
+  BOOST_TEST(
+      std::isnan(mock_step->GetBdaBuffers()[0]->GetRows()[0].data->real()));
```

Comment:
Puttng mock_step->GetBdaBuffers()[0]->GetRows()[0].data in a variable may increase readability a bit.